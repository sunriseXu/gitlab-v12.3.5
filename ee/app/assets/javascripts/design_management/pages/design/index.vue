<script>
import Mousetrap from 'mousetrap';
import createFlash from '~/flash';
import { s__ } from '~/locale';
import { GlLoadingIcon } from '@gitlab/ui';
import allVersionsMixin from '../../mixins/all_versions';
import Toolbar from '../../components/toolbar/index.vue';
import DesignImage from '../../components/image.vue';
import DesignOverlay from '../../components/design_overlay.vue';
import DesignDiscussion from '../../components/design_notes/design_discussion.vue';
import DesignReplyForm from '../../components/design_notes/design_reply_form.vue';
import getDesignQuery from '../../graphql/queries/getDesign.query.graphql';
import createImageDiffNoteMutation from '../../graphql/mutations/createImageDiffNote.mutation.graphql';
import { extractDiscussions } from '../../utils/design_management_utils';

export default {
  components: {
    DesignImage,
    DesignOverlay,
    DesignDiscussion,
    Toolbar,
    DesignReplyForm,
    GlLoadingIcon,
  },
  mixins: [allVersionsMixin],
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      design: {},
      comment: '',
      annotationCoordinates: null,
      overlayDimensions: {
        width: 0,
        height: 0,
      },
      projectPath: '',
      isNoteSaving: false,
    };
  },
  apollo: {
    design: {
      query: getDesignQuery,
      variables() {
        return {
          id: this.id,
          version: this.designsVersion,
        };
      },
      result({ data }) {
        if (!data) {
          createFlash(s__('DesignManagement|Could not find design, please try again.'));
          this.$router.push({ name: 'designs' });
        }
        if (this.$route.query.version && !this.hasValidVersion) {
          createFlash(s__('DesignManagement|Requested design version does not exist'));
          this.$router.push({ name: 'designs' });
        }
      },
    },
  },
  computed: {
    isLoading() {
      return this.$apollo.queries.design.loading;
    },
    discussions() {
      return extractDiscussions(this.design.discussions);
    },
    discussionStartingNotes() {
      return this.discussions.map(discussion => discussion.notes[0]);
    },
    markdownPreviewPath() {
      return `/${this.projectPath}/preview_markdown?target_type=Issue`;
    },
    isSubmitButtonDisabled() {
      return this.comment.trim().length === 0;
    },
    renderDiscussions() {
      return this.discussions.length || this.annotationCoordinates;
    },
  },
  mounted() {
    Mousetrap.bind('esc', this.closeDesign);
  },
  beforeDestroy() {
    Mousetrap.unbind('esc', this.closeDesign);
  },
  methods: {
    addImageDiffNote() {
      const { x, y, width, height } = this.annotationCoordinates;
      this.isNoteSaving = true;
      return this.$apollo
        .mutate({
          mutation: createImageDiffNoteMutation,
          variables: {
            input: {
              noteableId: this.design.id,
              body: this.comment,
              position: {
                headSha: this.design.diffRefs.headSha,
                baseSha: this.design.diffRefs.baseSha,
                startSha: this.design.diffRefs.startSha,
                x,
                y,
                width,
                height,
                paths: {
                  newPath: this.design.fullPath,
                },
              },
            },
          },
          update: (store, { data: { createImageDiffNote } }) => {
            const data = store.readQuery({
              query: getDesignQuery,
              variables: {
                id: this.id,
                version: this.designsVersion,
              },
            });
            const newDiscussion = {
              __typename: 'DiscussionEdge',
              node: {
                // False positive i18n lint: https://gitlab.com/gitlab-org/frontend/eslint-plugin-i18n/issues/26
                // eslint-disable-next-line @gitlab/i18n/no-non-i18n-strings
                __typename: 'Discussion',
                id: createImageDiffNote.note.discussion.id,
                replyId: createImageDiffNote.note.discussion.replyId,
                notes: {
                  __typename: 'NoteConnection',
                  edges: [
                    {
                      __typename: 'NoteEdge',
                      node: createImageDiffNote.note,
                    },
                  ],
                },
              },
            };
            data.design.discussions.edges.push(newDiscussion);
            store.writeQuery({ query: getDesignQuery, data });
          },
        })
        .then(() => {
          this.closeCommentForm();
          this.isNoteSaving = false;
        })
        .catch(e => {
          this.isNoteSaving = false;
          createFlash(s__('DesignManagement|Could not create new discussion, please try again.'));
          throw e;
        });
    },
    openCommentForm(position) {
      const { x, y } = position;
      const { width, height } = this.overlayDimensions;
      this.annotationCoordinates = {
        ...this.annotationCoordinates,
        x,
        y,
        width,
        height,
      };
    },
    closeCommentForm() {
      this.comment = '';
      this.annotationCoordinates = null;
    },
    setOverlayDimensions(position) {
      this.overlayDimensions.width = position.width;
      this.overlayDimensions.height = position.height;
    },
    closeDesign() {
      this.$router.push({
        name: 'designs',
        query: this.$route.query,
      });
    },
  },
  beforeRouteUpdate(to, from, next) {
    this.closeCommentForm();
    next();
  },
};
</script>

<template>
  <div class="design-detail fixed-top w-100 position-bottom-0 d-sm-flex justify-content-center">
    <gl-loading-icon v-if="isLoading" size="xl" class="align-self-center" />
    <template v-else>
      <div class="d-flex flex-column w-100">
        <toolbar
          :id="id"
          :name="design.filename"
          :updated-at="design.updatedAt"
          :updated-by="design.updatedBy"
        />
        <div class="d-flex flex-column w-100 h-100 mh-100 position-relative">
          <design-image
            :image="design.image"
            :name="design.filename"
            @setOverlayDimensions="setOverlayDimensions"
          />
          <design-overlay
            :position="overlayDimensions"
            :notes="discussionStartingNotes"
            :current-comment-form="annotationCoordinates"
            @openCommentForm="openCommentForm"
          />
        </div>
      </div>
      <div class="image-notes">
        <template v-if="renderDiscussions">
          <design-discussion
            v-for="(discussion, index) in discussions"
            :key="discussion.id"
            :discussion="discussion"
            :design-id="id"
            :noteable-id="design.id"
            :discussion-index="index + 1"
            :markdown-preview-path="markdownPreviewPath"
          />
          <design-reply-form
            v-if="annotationCoordinates"
            v-model="comment"
            :is-saving="isNoteSaving"
            :markdown-preview-path="markdownPreviewPath"
            @submitForm="addImageDiffNote"
            @cancelForm="closeCommentForm"
          />
        </template>
        <h2 v-else class="new-discussion-disclaimer m-0">
          {{ __("Click the image where you'd like to start a new discussion") }}
        </h2>
      </div>
    </template>
  </div>
</template>
