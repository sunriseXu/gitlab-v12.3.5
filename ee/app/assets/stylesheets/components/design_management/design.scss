.design-image {
  object-fit: contain;
}

.design-detail {
  overflow-y: scroll;
}

.design-list-item .design-event {
  top: $gl-padding;
  right: $gl-padding;
}

.image-notes {
  overflow-y: scroll;
  padding: 0 $gl-padding;
  padding-top: 50px;
  background-color: $white-light;
  min-width: 400px;

  li {
    list-style: none;
  }

  .badge.badge-pill {
    margin-left: $gl-padding;
    background-color: $blue-400;
    color: $white-light;
    border: $white-light 1px solid;
    min-height: 28px;
    padding: 7px 10px;
    border-radius: $gl-padding;
  }

  .design-discussion {
    padding: $gl-padding 0;
    margin: $gl-padding 0;

    &::before {
      content: '';
      border-left: 1px solid $gray-200;
      position: absolute;
      left: 28px;
      top: -18px;
      height: 18px;
    }

    .design-note {
      padding: $gl-padding;

      a {
        color: inherit;
      }
    }

    .reply-wrapper {
      padding: $gl-padding;
      padding-bottom: 0;
    }
  }

  .reply-wrapper {
    border-top: 1px solid $border-color;
  }

  .new-discussion-disclaimer {
    line-height: 20px;
  }
}

@media (max-width: map-get($grid-breakpoints, sm)) {
  .design-detail {
    overflow-y: scroll;
  }

  .image-notes {
    overflow-y: auto;
    min-width: 100%;
  }
}
