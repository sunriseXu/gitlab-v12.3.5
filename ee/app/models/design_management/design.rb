# frozen_string_literal: true

module DesignManagement
  class Design < ApplicationRecord
    include Noteable
    include Gitlab::FileTypeDetection
    include Gitlab::Utils::StrongMemoize
    include Referable

    belongs_to :project, inverse_of: :designs
    belongs_to :issue

    has_many :actions
    has_many :versions, through: :actions, class_name: 'DesignManagement::Version', inverse_of: :designs
    # This is a polymorphic association, so we can't count on FK's to delete the
    # data
    has_many :notes, as: :noteable, dependent: :delete_all # rubocop:disable Cop/ActiveRecordDependent

    validates :project, :issue, :filename, presence: true
    validates :filename, uniqueness: { scope: :issue_id }
    validate :validate_file_is_image

    alias_attribute :title, :filename

    # Find designs visible at the given version
    #
    # @param version [nil, DesignManagement::Version]:
    #   the version at which the designs must be visible
    #   Passing `nil` is the same as passing the most current version
    #
    # Restricts to designs
    # - created at least *before* the given version
    # - not deleted as of the given version.
    #
    # As a query, we ascertain this by finding the last event prior to
    # (or equal to) the cut-off, and seeing whether that version was a deletion.
    scope :visible_at_version, -> (version) do
      deletion = ::DesignManagement::Action.events[:deletion]
      designs = arel_table
      actions = ::DesignManagement::Action
        .most_recent.up_to_version(version)
        .arel.as('most_recent_actions')

      join = designs.join(actions)
        .on(actions[:design_id].eq(designs[:id]))

      joins(join.join_sources).where(actions[:event].not_eq(deletion))
    end

    # A design is current if the most recent event is not a deletion
    scope :current, -> { visible_at_version(nil) }

    def status
      if new_design?
        :new
      elsif deleted?
        :deleted
      else
        :current
      end
    end

    def deleted?
      most_recent_action&.deletion?
    end

    def most_recent_action
      strong_memoize(:most_recent_action) { actions.ordered.last }
    end

    # A reference for a design is the issue reference, indexed by the filename
    # with an optional infix when full.
    #
    # e.g.
    #   #123[homescreen.png]
    #   other-project#72[sidebar.jpg]
    #   #38/designs[transition.gif]
    def to_reference(from = nil, full: false)
      infix = full ? '/designs' : ''

      "%s%s[%s]" % [issue.to_reference(from, full: full), infix, filename]
    end

    def to_ability_name
      'design'
    end

    def description
      ''
    end

    def new_design?
      strong_memoize(:new_design) { actions.none? }
    end

    def full_path
      @full_path ||= File.join(DesignManagement.designs_directory, "issue-#{issue.iid}", filename)
    end

    def diff_refs
      strong_memoize(:diff_refs) do
        head_version.presence && repository.commit(head_version.sha).diff_refs
      end
    end

    def clear_version_cache
      [versions, actions].each(&:reset)
      [:new_design, :diff_refs, :head_sha, :most_recent_action].each do |key|
        clear_memoization(key)
      end
    end

    def repository
      project.design_repository
    end

    def user_notes_count
      user_notes_count_service.count
    end

    def after_note_changed(note)
      user_notes_count_service.delete_cache unless note.system?
    end
    alias_method :after_note_created,   :after_note_changed
    alias_method :after_note_destroyed, :after_note_changed

    private

    def head_version
      strong_memoize(:head_sha) { versions.ordered.first }
    end

    def validate_file_is_image
      unless image?
        message = _("Only these extensions are supported: %{extension_list}") % {
          extension_list: Gitlab::FileTypeDetection::IMAGE_EXT.join(", ")
        }
        errors.add(:filename, message)
      end
    end

    def user_notes_count_service
      strong_memoize(:user_notes_count_service) do
        DesignManagement::DesignUserNotesCountService.new(self) # rubocop: disable CodeReuse/ServiceClass
      end
    end
  end
end
