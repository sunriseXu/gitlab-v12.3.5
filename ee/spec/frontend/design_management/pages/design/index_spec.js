import { shallowMount } from '@vue/test-utils';
import DesignIndex from 'ee/design_management/pages/design/index.vue';
import DesignDiscussion from 'ee/design_management/components/design_notes/design_discussion.vue';
import DesignReplyForm from 'ee/design_management/components/design_notes/design_reply_form.vue';
import createImageDiffNoteMutation from 'ee/design_management/graphql/mutations/createImageDiffNote.mutation.graphql';
import design from '../../mock_data/design';

jest.mock('mousetrap', () => ({
  bind: jest.fn(),
  unbind: jest.fn(),
}));

describe('Design management design index page', () => {
  let wrapper;
  const newComment = 'new comment';
  const annotationCoordinates = {
    x: 10,
    y: 10,
    width: 100,
    height: 100,
  };
  const mutationVariables = {
    mutation: createImageDiffNoteMutation,
    update: expect.anything(),
    variables: {
      input: {
        body: newComment,
        noteableId: design.id,
        position: {
          headSha: 'headSha',
          baseSha: 'baseSha',
          startSha: 'startSha',
          paths: {
            newPath: 'full-design-path',
          },
          ...annotationCoordinates,
        },
      },
    },
  };
  const mutate = jest.fn(() => Promise.resolve());

  const findDiscussions = () => wrapper.findAll(DesignDiscussion);
  const findDiscussionForm = () => wrapper.find(DesignReplyForm);

  function createComponent(loading = false) {
    const $apollo = {
      queries: {
        design: {
          loading,
        },
      },
      mutate,
    };

    wrapper = shallowMount(DesignIndex, {
      sync: false,
      propsData: { id: '1' },
      mocks: { $apollo },
    });
  }

  function setDesign() {
    createComponent(true);
    wrapper.vm.$apollo.queries.design.loading = false;
  }

  afterEach(() => {
    wrapper.destroy();
  });

  it('sets loading state', () => {
    createComponent(true);

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders design index', () => {
    setDesign();

    wrapper.setData({
      design,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('when has no discussions', () => {
    beforeEach(() => {
      setDesign();

      wrapper.setData({
        design: {
          ...design,
          discussions: {
            edges: [],
          },
        },
      });
    });

    it('does not render discussions', () => {
      expect(findDiscussions().exists()).toBe(false);
    });

    it('renders a message about possibility to create a new discussion', () => {
      expect(wrapper.find('.new-discussion-disclaimer').exists()).toBe(true);
    });
  });

  describe('when has discussions', () => {
    beforeEach(() => {
      setDesign();

      wrapper.setData({
        design,
      });
    });

    it('renders correct amount of discussions', () => {
      expect(wrapper.findAll(DesignDiscussion).length).toBe(1);
    });
  });

  it('opens a new discussion form', () => {
    setDesign();

    wrapper.setData({
      design: {
        ...design,
        discussions: {
          edges: [],
        },
      },
    });

    wrapper.vm.openCommentForm({ x: 0, y: 0 });

    wrapper.vm.$nextTick(() => {
      expect(findDiscussionForm().exists()).toBe(true);
    });
  });

  it('sends a mutation on submitting form and closes form', () => {
    setDesign();

    wrapper.setData({
      design: {
        ...design,
        discussions: {
          edges: [],
        },
      },
      annotationCoordinates,
      comment: newComment,
    });

    wrapper.vm.$nextTick(() => {
      findDiscussionForm().vm.$emit('submitForm');

      expect(mutate).toHaveBeenCalledWith(mutationVariables);
      const addNote = wrapper.vm.addImageDiffNote();

      return addNote.then(() => {
        expect(findDiscussionForm().exists()).toBe(false);
      });
    });
  });
});
