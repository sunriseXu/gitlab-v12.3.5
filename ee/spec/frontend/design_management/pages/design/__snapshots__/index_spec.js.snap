// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design management design index page renders design index 1`] = `
<div
  class="design-detail fixed-top w-100 position-bottom-0 d-sm-flex justify-content-center"
>
  <glloadingicon-stub
    class="align-self-center"
    color="orange"
    label="Loading"
    size="xl"
  />
</div>
`;

exports[`Design management design index page sets loading state 1`] = `
<div
  class="design-detail fixed-top w-100 position-bottom-0 d-sm-flex justify-content-center"
>
  <glloadingicon-stub
    class="align-self-center"
    color="orange"
    label="Loading"
    size="xl"
  />
</div>
`;
